File: ios/joinery-kit/Sources/JoineryAIChatKit/ChatSettingsSheet.swift
import SwiftUI
import JoineryKit

/// Per-chat settings: model, capability toggles, reasoning effort, and sampling
/// / instructions. Picker and toggle changes apply immediately (persisted for
/// an existing chat, seeded onto a new chat's first send); the free-text fields
/// commit on Done. The same controls and validator the web status strip uses.
struct ChatSettingsSheet: View {
    @ObservedObject var store: ChatThreadStore
    @Environment(\.dismiss) private var dismiss

    // Free-text fields commit on Done rather than each keystroke.
    @State private var temperature = ""
    @State private var topP = ""
    @State private var maxTokens = ""
    @State private var instructions = ""

    private let thinkingLevels = ["off", "low", "medium", "high"]

    var body: some View {
        NavigationStack {
            Group {
                if let meta = store.meta {
                    form(meta)
                } else {
                    ProgressView()
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                        .accessibilityIdentifier("chat_settings_loading")
                }
            }
            .navigationTitle("Chat settings")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .confirmationAction) {
                    Button("Done") { commitText(); dismiss() }
                        .accessibilityIdentifier("chat_settings_done")
                }
            }
        }
        .task {
            await store.loadMeta()
            syncText()
        }
    }

    private func form(_ meta: ChatControlsMeta) -> some View {
        Form {
            Section {
                Picker("Model", selection: modelBinding) {
                    ForEach(meta.models) { model in
                        Text(model.label).tag(model.id)
                    }
                }
                .pickerStyle(.navigationLink)
                .accessibilityIdentifier("chat_set_model")
            } header: {
                Text("Model")
            } footer: {
                if !meta.isPrivate(store.controls.model) {
                    Text("This model isn't private — avoid sending sensitive personal data.")
                }
            }

            Section {
                Toggle("Data access", isOn: toggleBinding(
                    "data_access", get: { store.controls.dataAccess }, set: { $0.dataAccess = $1 }))
                    .accessibilityIdentifier("chat_set_data_access")
                Toggle("Web search", isOn: toggleBinding(
                    "web_search", get: { store.controls.webSearch }, set: { $0.webSearch = $1 }))
                    .disabled(!meta.webSearchAvailable)
                    .accessibilityIdentifier("chat_set_web_search")
            } header: {
                Text("Capabilities")
            } footer: {
                Text("Data access lets the assistant read your data and, with your confirmation, make changes.")
            }

            Section("Reasoning") {
                Picker("Thinking", selection: thinkingBinding) {
                    ForEach(thinkingLevels, id: \.self) { level in
                        Text(level.capitalized).tag(level)
                    }
                }
                .accessibilityIdentifier("chat_set_thinking")
            }

            Section {
                numberField("Temperature", text: $temperature, placeholder: meta.defaults.temperature, keyboard: .decimalPad)
                numberField("Top-p", text: $topP, placeholder: meta.defaults.topP, keyboard: .decimalPad)
                numberField("Max tokens", text: $maxTokens, placeholder: meta.defaults.maxTokens, keyboard: .numberPad)
            } header: {
                Text("Sampling")
            } footer: {
                Text("Leave blank to use the default.")
            }

            Section("Instructions") {
                TextField("Custom instructions (optional)", text: $instructions, axis: .vertical)
                    .lineLimit(3...8)
                    .accessibilityIdentifier("chat_set_instructions")
            }
        }
    }

    private func numberField(_ label: String, text: Binding<String>, placeholder: String, keyboard: UIKeyboardType) -> some View {
        HStack {
            Text(label)
            Spacer()
            TextField(placeholder.isEmpty ? "default" : placeholder, text: text)
                .keyboardType(keyboard)
                .multilineTextAlignment(.trailing)
                .frame(maxWidth: 120)
        }
    }

    // MARK: Bindings (immediate commit)

    private var modelBinding: Binding<String> {
        Binding(
            get: { store.controls.model.isEmpty ? (store.meta?.defaults.model ?? "") : store.controls.model },
            set: { value in store.setControl(field: "model", value: value) { $0.model = value } }
        )
    }

    private var thinkingBinding: Binding<String> {
        Binding(
            get: { store.controls.thinkingLevel },
            set: { value in store.setControl(field: "thinking_level", value: value) { $0.thinkingLevel = value } }
        )
    }

    private func toggleBinding(_ field: String, get: @escaping () -> Bool,
                               set: @escaping (inout ChatControlValues, Bool) -> Void) -> Binding<Bool> {
        Binding(
            get: get,
            set: { on in store.setControl(field: field, value: on ? "1" : "0") { set(&$0, on) } }
        )
    }

    // MARK: Free-text commit

    private func syncText() {
        temperature = store.controls.temperature
        topP = store.controls.topP
        maxTokens = store.controls.maxTokens
        instructions = store.controls.instructions
    }

    private func commitText() {
        commit("temperature", temperature, current: store.controls.temperature) { $0.temperature = temperature }
        commit("top_p", topP, current: store.controls.topP) { $0.topP = topP }
        commit("max_tokens", maxTokens, current: store.controls.maxTokens) { $0.maxTokens = maxTokens }
        commit("instructions", instructions, current: store.controls.instructions) { $0.instructions = instructions }
    }

    private func commit(_ field: String, _ value: String, current: String,
                        _ apply: @escaping (inout ChatControlValues) -> Void) {
        guard value != current else { return }
        store.setControl(field: field, value: value, apply: apply)
    }
}

File: android/joinery-android-ai-chat/src/main/java/com/getjoinery/aichat/ChatSettingsSheet.kt
package com.getjoinery.aichat

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Close
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ExposedDropdownMenuBox
import androidx.compose.material3.ExposedDropdownMenuDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties

/**
 * Per-chat settings: model, capability toggles, reasoning effort, and
 * sampling / instructions. Picker and toggle changes apply immediately
 * (persisted for an existing chat, seeded onto a new chat's first send); the
 * free-text fields commit on Done. The same controls and validator the web
 * status strip uses.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ChatSettingsSheet(store: ChatThreadStore, onDismiss: () -> Unit) {
    // Free-text fields commit on Done rather than each keystroke.
    var temperature by remember { mutableStateOf("") }
    var topP by remember { mutableStateOf("") }
    var maxTokens by remember { mutableStateOf("") }
    var instructions by remember { mutableStateOf("") }
    var synced by remember { mutableStateOf(false) }

    LaunchedEffect(Unit) {
        store.loadMeta()
        if (!synced) {
            temperature = store.controls.temperature
            topP = store.controls.topP
            maxTokens = store.controls.maxTokens
            instructions = store.controls.instructions
            synced = true
        }
    }

    fun commitText() {
        if (temperature != store.controls.temperature) {
            store.setControl("temperature", temperature) { it.copy(temperature = temperature) }
        }
        if (topP != store.controls.topP) {
            store.setControl("top_p", topP) { it.copy(topP = topP) }
        }
        if (maxTokens != store.controls.maxTokens) {
            store.setControl("max_tokens", maxTokens) { it.copy(maxTokens = maxTokens) }
        }
        if (instructions != store.controls.instructions) {
            store.setControl("instructions", instructions) { it.copy(instructions = instructions) }
        }
    }

    Dialog(
        onDismissRequest = { commitText(); onDismiss() },
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        Surface(Modifier.fillMaxSize()) {
            Scaffold(topBar = {
                TopAppBar(
                    title = { Text("Chat settings") },
                    navigationIcon = {
                        IconButton(onClick = { commitText(); onDismiss() }) {
                            Icon(Icons.Outlined.Close, contentDescription = "Close")
                        }
                    },
                    actions = {
                        TextButton(
                            onClick = { commitText(); onDismiss() },
                            modifier = Modifier.testTag("chat_settings_done"),
                        ) { Text("Done") }
                    },
                )
            }) { padding ->
                val meta = store.meta
                if (meta == null) {
                    Box(Modifier.fillMaxSize().padding(padding), contentAlignment = Alignment.Center) {
                        CircularProgressIndicator(Modifier.testTag("chat_settings_loading"))
                    }
                } else {
                    Column(
                        Modifier
                            .fillMaxSize()
                            .padding(padding)
                            .verticalScroll(rememberScrollState())
                            .padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(12.dp),
                    ) {
                        SettingsDropdown(
                            label = "Model",
                            options = meta.models.map { it.id to it.label },
                            selected = store.controls.model.ifEmpty { meta.defaults.model },
                            tag = "chat_set_model",
                        ) { value ->
                            store.setControl("model", value) { it.copy(model = value) }
                        }
                        if (!meta.isPrivate(store.controls.model)) {
                            Text(
                                "This model isn't private — avoid sending sensitive personal data.",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                        HorizontalDivider()
                        SettingsSwitch(
                            label = "Data access",
                            checked = store.controls.dataAccess,
                            tag = "chat_set_data_access",
                        ) { on ->
                            store.setControl("data_access", if (on) "1" else "0") { it.copy(dataAccess = on) }
                        }
                        SettingsSwitch(
                            label = "Web search",
                            checked = store.controls.webSearch,
                            enabled = meta.webSearchAvailable,
                            tag = "chat_set_web_search",
                        ) { on ->
                            store.setControl("web_search", if (on) "1" else "0") { it.copy(webSearch = on) }
                        }
                        Text(
                            "Data access lets the assistant read your data and, with your confirmation, make changes.",
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        HorizontalDivider()
                        SettingsDropdown(
                            label = "Thinking",
                            options = listOf("off", "low", "medium", "high").map {
                                it to it.replaceFirstChar { c -> c.uppercase() }
                            },
                            selected = store.controls.thinkingLevel,
                            tag = "chat_set_thinking",
                        ) { value ->
                            store.setControl("thinking_level", value) { it.copy(thinkingLevel = value) }
                        }
                        HorizontalDivider()
                        NumberField("Temperature", temperature, meta.defaults.temperature) { temperature = it }
                        NumberField("Top-p", topP, meta.defaults.topP) { topP = it }
                        NumberField("Max tokens", maxTokens, meta.defaults.maxTokens) { maxTokens = it }
                        Text(
                            "Leave blank to use the default.",
                            style = MaterialTheme.typography.labelMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        HorizontalDivider()
                        OutlinedTextField(
                            value = instructions,
                            onValueChange = { instructions = it },
                            label = { Text("Custom instructions (optional)") },
                            minLines = 3,
                            maxLines = 8,
                            modifier = Modifier.fillMaxWidth().testTag("chat_set_instructions"),
                        )
                    }
                }
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun SettingsDropdown(
    label: String,
    options: List<Pair<String, String>>,
    selected: String,
    tag: String,
    onSelect: (String) -> Unit,
) {
    var open by remember { mutableStateOf(false) }
    val current = options.firstOrNull { it.first == selected }?.second ?: selected

    ExposedDropdownMenuBox(expanded = open, onExpandedChange = { open = it }) {
        OutlinedTextField(
            value = current,
            onValueChange = {},
            readOnly = true,
            label = { Text(label) },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = open) },
            modifier = Modifier.fillMaxWidth().menuAnchor().testTag(tag),
        )
        ExposedDropdownMenu(expanded = open, onDismissRequest = { open = false }) {
            options.forEach { (value, text) ->
                DropdownMenuItem(
                    text = { Text(text) },
                    onClick = {
                        open = false
                        onSelect(value)
                    },
                )
            }
        }
    }
}

@Composable
private fun SettingsSwitch(
    label: String,
    checked: Boolean,
    enabled: Boolean = true,
    tag: String,
    onChange: (Boolean) -> Unit,
) {
    Row(
        Modifier.fillMaxWidth().testTag(tag),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(label, modifier = Modifier.weight(1f))
        Switch(checked = checked, onCheckedChange = onChange, enabled = enabled)
    }
}

@Composable
private fun NumberField(
    label: String,
    value: String,
    placeholder: String,
    onChange: (String) -> Unit,
) {
    OutlinedTextField(
        value = value,
        onValueChange = onChange,
        label = { Text(label) },
        placeholder = { Text(placeholder.ifEmpty { "default" }) },
        singleLine = true,
        keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
        modifier = Modifier.fillMaxWidth(),
    )
}
